### PR TITLE
Update and correct pow function for PauliSum

### PR DESCRIPTION
**Pull request by TarunSinghania**
Updated pow function for PauliSum to use binary exponentiation.

### CODE REVIEW

File: cirq-core/cirq/ops/linear_combinations.py [hunk 1]
```diff
@@ -839,10 +839,13 @@ def __pow__(self, exponent: int):
...
         if exponent > 0:
-            base = self.copy()
-            for _ in range(exponent - 1):
-                base *= base
-            return base
+            remainder = PauliSum.from_pauli_strings([PauliString(coefficient=1)])
+            while exponent > 0 :
+                if exponent & 1 :
+                    remainder = remainder * self
+                self = self * self
```
Comment (tanujkhattar) [inline cirq-core/cirq/ops/linear_combinations.py:846]: You should not change `self`. Instead make a copy of `self` and use that for exponentiation.

**Review by tanujkhattar — CHANGES_REQUESTED** [review]
Please add a new test to cirq-core/cirq/ops/linear_combinations_test.py which covers this change, i.e. the test should have failed on the old code and pass on new code.

**Comment by tanujkhattar** [comment]
closes https://github.com/quantumlib/Cirq/issues/6017

File: cirq-core/cirq/ops/linear_combinations.py (hunk 1 above)
Comment (TarunSinghania) [reply cirq-core/cirq/ops/linear_combinations.py]: thanks for the review, I have modified the code to use a copy.

File: cirq-core/cirq/ops/linear_combinations.py [hunk 2]
```diff
@@ -839,10 +839,14 @@ def __pow__(self, exponent: int):
...
         if exponent > 0:
+            remainder = PauliSum.from_pauli_strings([PauliString(coefficient=1)])
             base = self.copy()
-            for _ in range(exponent - 1):
-                base *= base
-            return base
+            while exponent > 0:
+                if exponent & 1:
+                    remainder = remainder * base
+                base = base * base
```
Comment (pavoljuhas) [inline cirq-core/cirq/ops/linear_combinations.py:847]: the last `base` value in the loop would be unused.
can you please move the `exponent` check before the base squaring?

**Comment by TarunSinghania** [comment]
Thanks everyone for your inputs, It was helpful in developing the below understanding:

There are two ways to implement the pow function i.e either use a linear multiplication approach or binary exponentiation. 

To understand the performance aspect of both of them. I am leaving two comments. One is for those who want a quick revision on PauliSum and its implementation of multiplication. The second is performance analysis and reason to choose linear multiplication.

**Comment by TarunSinghania** [comment]
__PauliSum__:

PauliSum is a linear combination of PauliStrings.  The way PauliSum stores it is as LinearDict[PauliString] = coefficient. The LinearDict implementation can be found in linear_dict.py

Multiplication: PauliSum*PauliSum : `__mul__` -> `__imul__` 
Flow of multiplication

Post Multiplication the number of terms can range from [1,N*M]

__PauliString__:

Tensor product of single qubit (non identity) pauli operations, each acting on a different qubit.If more than one pauli operation acts on the same set of qubits, their composition is immediately reduced to an equivalent (possibly multi-qubit) Pauli operator

The way we store PauliStrings are `Dict[TKey, 'cirq.Pauli']`, where Tkey = TypeVar('TKey', bound=raw_types.Qid) i.e a quantum object (qubit, qudit, resonator) and Where cirq.Pauli is a pauli gate operation which can be X,Y, Z Gates. 

Multiplication of two PauliStrings: PauliString1*PauliString2 
`__mul__ -> constructor -> _inplace_left_multiply_by -> _imul_helper_checkpoint -> _imul_helper -> _imul_atom_helper`

Pauli String Multiplication

if we multiply two paulistring with length a and b respectively, we will be left with a string of length [1,a+b] depending on the underlying objects which make up the Paulistrings

The time complexity of the multiplication operation is o(a + b) in average case considering pop is o(1) 

Summary:Two PauliSum with N pauli strings and M pauliStrings take 
`O(N*M)*(Cost of multiplying PauliStrings)` where cost of multiplying two PauliStrings of length a and b is O(a+b)

**Comment by TarunSinghania** [comment]
__Various Approaches:__

If we have a PauliSum with n PauliStrings each of length l acting on x qubits. Time complexity to exponentiate it to the power e

__Linear Multiplication__
```
def __pow__(self, exponent: int):
       if not isinstance(exponent, numbers.Integral):
           return NotImplemented
       if exponent == 0:
           return PauliSum(value.LinearDict({frozenset(): 1 + 0j}))
       if exponent > 0:
           base = self.copy()
           for _ in range(exponent - 1):
               base *= self.copy()
           return base
       return NotImplemented
 ```

Time Complexity: `n + n**2 + n**3 + …n***e`

__Binary exponentiation__
```
def __pow__(self, exponent: int):
       if not isinstance(exponent, numbers.Integral):
           return NotImplemented
       if exponent == 0:
           return PauliSum(value.LinearDict({frozenset(): 1 + 0j}))
       if exponent > 0:
           remainder = PauliSum.from_pauli_strings([PauliString(coefficient=1)])
	    base = self.copy()
           while exponent > 0 :
               if exponent&1 :
                   remainder = remainder * base
               base *= base
               exponent = exponent >> 1
           return remainder
       return NotImplemented
```
Comparing Linear and Binary exponentiation

Exponent e | Linear | Binary exponentiation
-- | -- | --
2 | n + n*\*2 | n + n*\*2 + n*\*2 + n*\*4
3 | <span> n + n*\*2 + n*\*3  </span> | n + n + n*\*2 + n*\*3 + n*\*4
4 | n + n*\*2 + n*\*3 + n*\*4 | n + n*\*2 + n*\*4 + n*\*4 + n*\*16
5 | n + n*\*2 + n*\*3 + n*\*4 + n*\*5 | n + n + n*\*2 + n*\*4 + n*\*5 + n*\*16

The reason why this performs worse than linear multiplication: we go in the loop when e reduces to 1 and perform a extra base*base which costs very high
In order to prevent this the algorithm can be modified as below:

__Modified Binary exponentiation 1__ 
```
def __pow__(self, exponent: int):
       if not isinstance(exponent, numbers.Integral):
           return NotImplemented
       if exponent == 0:
           return PauliSum(value.LinearDict({frozenset(): 1 + 0j}))
       if exponent > 1:
           remainder = PauliSum.from_pauli_strings([PauliString(coefficient=1)])
	    base = self.copy()
           while exponent > 1 :
               if exponent&1 :
                   remainder = remainder * base
		exponent = exponent >> 1
               base *= base
	   return remainder * base     
       return NotImplemented
```
Exponent e | Linear | Binary exponentiation
-- | -- | --
2 | n + n*\*2 | n + n*\*2 + n*\*2 + n*\*4
3 | ‘n + n*\*2 + n*\*3’ | n + n + n*\*2 + n*\*3 + n*\*4
4 | n + n*\*2 + n*\*3 + n*\*4 | n + n*\*2 + n*\*4 + n*\*4 + n*\*16
5 | n + n*\*2 + n*\*3 + n*\*4 + n*\*5 | n + n + n*\*2  + n*\*4 + n*\*5 + n*\*16

We are still multiplying if remainder equals identity when we exit the loop,  which can be removed

__Modified Binary exponentiation 2__ 
```
def __pow__(self, exponent: int):
       if not isinstance(exponent, numbers.Integral):
           return NotImplemented
       if exponent == 0:
           return PauliSum(value.LinearDict({frozenset(): 1 + 0j}))
       if exponent > 1:
           remainder = PauliSum.from_pauli_strings([PauliString(coefficient=1)])
	    base = self.copy()
           while exponent > 1 :
               if exponent&1 :
                   remainder = remainder * base
		exponent = exponent >> 1
               base *= base
	   return remainder * base     
       return NotImplemented
```
Exponent e | Linear | Binary exponentiation
-- | -- | --
2 | n + n*\*2 | n + n*\*2 + n*\*2
3 | n + n*\*2 + n*\*3 | n + n + n*\*2 + n*\*3
4 | n + n*\*2 + n*\*3 + n*\*4 | n + n*\*2 + n*\*4 + n*\*4
5 | n + n*\*2 + n*\*3 + n*\*4 + n*\*5 | n + n + n*\*2 + n*\*4 + n*\*5

We are still multiplying if remainder equals identity when we exit the loop,  which can be removed

__Modified Binary exponentiation 3__ 
```
def __pow__(self, exponent: int):
       if not isinstance(exponent, numbers.Integral):
           return NotImplemented
       if exponent == 0:
           return PauliSum(value.LinearDict({frozenset(): 1 + 0j}))
       if exponent > 1:
           remainder = PauliSum.from_pauli_strings([PauliString(coefficient=1)])
	    base = self.copy()
	    identity = remainder.copy()
           while exponent > 1 :
               if exponent&1 :
                   remainder = remainder * base
		exponent = exponent >> 1
               base*=base
	   if(remainder != identity) base*=remainder
	   return base     
       return NotImplemented
```

 e | Linear | Binary exponentiation
-- | -- | --
2 | n + n*\*2 | n + n*\*2
3 | n + n*\*2 + n*\*3 | n + n + n*\*2 + n*\*3
4 | n + n*\*2 + n*\*3 + n*\*4 | n + n*\*2 + n*\*4
5 | n + n*\*2 + n*\*3 + n*\*4 + n*\*5 | n + n + n*\*2  + n*\*4 + n*\*5
6 | n + n*\*2 + n*\*3 + n*\*4 + n*\*5 + n*\*6 | n + n*\*2 + n*\*2 + n*\*4 +n*\*6
7 | n + n*\*2 + n*\*3 + n*\*4 + n*\*5 + n*\*6 + n*\*7 | n + n + n*\*2 + n*\*3 + n*\*4 + n*\*7

Observations:
We can see from e = 7 for example we are preventing n**5 + n**6 - n calculations if we opt for binary exponentiation. Even though the number of iterations inside the loop is reduced to log(n) each iteration the cost of multiplication grows exponentially. The improvements for low e values (2-20) and various values of n are indicated in the attached sheet
For lower values of N (2-10) the improvements can go upto 40-50% for certain values of e.
As N increases the improvements reduce to 6% for n = 16 and 1-2% for n= 50
The only place where the binary method performs bad w.r.t linear is when e equals 3 where we do extra n steps. This can be handled as  for n <=3 
We can apply regular linear multiplication and for n>3 binary exponentiaiton. 

__Final Binary exponentiation__ 
```
def __pow__(self, exponent: int):
       if not isinstance(exponent, numbers.Integral):
           return NotImplemented
       identity = PauliSum.from_pauli_strings([PauliString(coefficient=1)])
       remainder = identity.copy()
       base = self.copy()
       if exponent == 0:
           return PauliSum(value.LinearDict({frozenset(): 1 + 0j}))
       if exponent > 0 and exponent <= 3:
           for _ in range(exponent - 1):
              base *= self.copy()
           return base
       if exponent > 3:
           while exponent > 1:
               if exponent & 1:
                   remainder = remainder * base
               base *= base
               exponent = exponent >> 1
           if remainder != identity:
               base*=remainder
           return base
       return NotImplemented
```
In all of these we are missing that, every time we multiply PauliSum of n terms with itself - we have n PauliStrings multiplying with each other taking o(n*\*2) steps. The resultant PauliSum can have number of terms between [1, n*\*2 - n + 1]. As we exponentiate, more terms will merge to identity and the number of terms in PauliSum decreases

Later as we run tests we will observe its effects.

**Comment by TarunSinghania** [comment]
Example test: (One can vary n, e) 
N is the number of terms in a PauliSum and e is the exponent to which we are raising it
Our PauliSum is like 1 + X(q0) .... X(qn) [Why I choose this is explained later]
```
import cirq
import time
n = 7
e = 20
qubits = cirq.LineQubit.range(n)
pstrings = []
for qubit in qubits:
   pstrings.append(cirq.PauliString(1, cirq.X(qubit)))
psum = cirq.PauliSum.from_pauli_strings(pstrings)
for n in range (2, e):
   st = time.process_time()
   psum2 = (psum**n)
   et = time.process_time()
   res = et - st
   print(psum2)
   print(res)
   #print('CPU Execution time:', res, 'seconds')
 ```
 
From [test runs](https://docs.google.com/spreadsheets/d/14DSyLreGKhftLqQx1GFqpFCnDv71tNevepJ1sKrD5dg/edit?usp=sharing) we see, for n = 7 and e = 7 the binary exponentiation takes twice as time as linear which is counter intuitive as  n*\*5 + n*\*6 - n PauliString multiplications should have been prevented in the binary exponentiation approach.

e | Linear | Binary exponentiation
-- | -- | --
7 | n + n*\*2 + n*\*3 + n*\*4 + n*\*5 + n*\*6 + n*\*7 | n + n + n*\*2 + n*\*3 + n*\*4 + n*\*7

Looking at more details

For Binary 
Time inside while loop for first n*\*2 + n was 0.001870999999999956
n here has 7 paulistrings, after multiplication n*\*2 has 21 PauliStrings
Time inside while loop for n*\*4 + n*\*3 = 0.018809000000000076
n*\*4 now has 56 paulistrings while n**3 has 41 paulistrings
Final multiplication costs: 0.08067499999999983

Overall total multiplications performed = 7 + 7 + 49 + 21\*7 + 21\*21 + 56\*41  = 2947

For Linear

1st iteration in while loop was 0.0014419999999999433
n**2 has 21 terms ....
Total time = 0.05953900000000001

Overall total multiplications performed  = 7 + 7\*7 + 21\*7 + 41\*7 + 56\*7 + 62\*7 + 63\*7 + 64\*7 = 2205

The rate at which the number of PauliStrings that decreased with increasing power when multipled by constant number of Paulistrings gave better results.

Overall we observe:
pow fun behaviour

Until the exponent saturates(meeting point) the cost of binary exponentiation is higher.This is because:
In Binary:
self\*self : At each step where we multiply is getting exponentially expensive
	Both left and right have large number of terms with each term consisting of multiple qubits
In Linear:
	base\*self.copy(): self.copy() is fixed and doesn't grow. Each PualiString in self.copy() is also smaller. On every step the number of terms added to base increases but the rate of growth is slower.
	
If n = x  meeting point is generally 2*\*(x-1) for the above category of PauliSum. Question is can we say for sure that meeting point cant be behind 2**(x-1) for other inputs as well. If we can prove that it ensures most delayed saturation we can probably implement the pow function as
``` 
if e < 2 ** (n - 1)
  linear multiplication exponentiation 
else 
  binary exponentiation 
```
Intuitively, I chose each PauliString in the PauliSum referencing a unique qubit as it tends to delay thes saturation. For saturation we need  X**2 = Y**2 = Z**2 = -iXYZ = I

However, given that we dont have a formal proof and the improvements are for larger values of e which itself increases as n increases. The percentage of cases in which it performs betters looks small. Will be going ahead and pushing the implementation of regular multiplication.

File: cirq-core/cirq/ops/linear_combinations.py (hunk 2 above)
Comment (TarunSinghania) [reply cirq-core/cirq/ops/linear_combinations.py]: Thanks for the observations @pavoljuhas, worked on it in my latest submission.

File: cirq-core/cirq/ops/linear_combinations.py (hunk 1 above)
Comment (TarunSinghania) [reply cirq-core/cirq/ops/linear_combinations.py]: Added tests as well on the latest commit

File: cirq-core/cirq/ops/linear_combinations.py [hunk 3]
```diff
@@ -840,9 +840,10 @@ def __pow__(self, exponent: int):
             return PauliSum(value.LinearDict({frozenset(): 1 + 0j}))
         if exponent > 0:
             base = self.copy()
+            answer = self.copy() 
             for _ in range(exponent - 1):
-                base *= base
-            return base
+                answer *= base
```
Comment (tanujkhattar) [inline cirq-core/cirq/ops/linear_combinations.py:845]: Can we just do `answer *= self` and not have `base` ?

File: cirq-core/cirq/ops/linear_combinations.py [hunk 4]
```diff
@@ -839,10 +839,10 @@ def __pow__(self, exponent: int):
         if exponent == 0:
             return PauliSum(value.LinearDict({frozenset(): 1 + 0j}))
         if exponent > 0:
-            base = self.copy()
+            answer = self.copy()
             for _ in range(exponent - 1):
-                base *= base
-            return base
+                answer *= self
+            return answer
```
Comment (pavoljuhas) [inline cirq-core/cirq/ops/linear_combinations.py:845]: Nit: most other functions here return `result`.  Please rename `answer` --> `result` for the sake of consistency.

File: cirq-core/cirq/ops/linear_combinations_test.py [hunk 5]
```diff
@@ -1156,6 +1156,18 @@ def test_pauli_sum_pow():
     for psum in [psum1, psum2, psum3, psum4]:
         assert cirq.approx_eq(psum**0, identity)
 
+    # tests for exponents greater than two for both even and odd 
+    psum5 = cirq.Z(q0) * cirq.Z(q1) + cirq.Z(q2) + cirq.Z(q3)
+    assert psum5 * psum5 * psum5 * psum5 * psum5 == psum5**5 
+
+    psum6 = cirq.X(q0) * cirq.Y(q1) + cirq.Z(q2) + cirq.X(q3)
+    assert psum6 * psum6 * psum6 * psum6 * psum6 * psum6 * psum6 * psum6 == psum6**8
```
Comment (pavoljuhas) [inline cirq-core/cirq/ops/linear_combinations_test.py:1164]: There is no test of exponent 1.  Perhaps we can replace this with a loop to test exponents for `range(1, 9)`.

File: cirq-core/cirq/ops/linear_combinations_test.py [hunk 6]
```diff
@@ -1156,6 +1156,18 @@ def test_pauli_sum_pow():
     for psum in [psum1, psum2, psum3, psum4]:
         assert cirq.approx_eq(psum**0, identity)
 
+    # tests for exponents greater than two for both even and odd 
+    psum5 = cirq.Z(q0) * cirq.Z(q1) + cirq.Z(q2) + cirq.Z(q3)
+    assert psum5 * psum5 * psum5 * psum5 * psum5 == psum5**5
```
Comment (pavoljuhas) [inline cirq-core/cirq/ops/linear_combinations_test.py:1161]: nit - trailing blanks, please make sure `git diff --check master...HEAD` passes without complaints.

**Review by pavoljuhas — CHANGES_REQUESTED** [review]
Nearly there, but we should also test with exponent = 1.

File: cirq-core/cirq/ops/linear_combinations.py (hunk 4 above)
Comment (TarunSinghania) [reply cirq-core/cirq/ops/linear_combinations.py:845]: Thanks for letting me know, updated

File: cirq-core/cirq/ops/linear_combinations_test.py (hunk 6 above)
Comment (TarunSinghania) [reply cirq-core/cirq/ops/linear_combinations_test.py:1161]: Thanks, will do that from now on

File: cirq-core/cirq/ops/linear_combinations_test.py (hunk 5 above)
Comment (TarunSinghania) [reply cirq-core/cirq/ops/linear_combinations_test.py:1164]: Sure, Updating the test

**Comment by pavoljuhas** [comment]
Hi Tarun, when updating the PR, please add new changes as new commits on top of your branch rather than amending old commits, which may result in complicated commit graphs with duplicate original and amended commits.  It is also not necessary to merge-in the mainline master in every update - provided the master does not touch files worked on in this PR.

It is not a big deal as in the end we squash each PR to a single commit on master,
but while the PR is open it is easier to follow its progress with a simple commit history.

File: cirq-core/cirq/ops/linear_combinations_test.py (hunk 6 above)
Comment (pavoljuhas) [reply cirq-core/cirq/ops/linear_combinations_test.py:1161]: The change is not in effect, please see https://github.com/quantumlib/Cirq/pull/6019#issuecomment-1468971247

File: cirq-core/cirq/ops/linear_combinations.py (hunk 4 above)
Comment (pavoljuhas) [reply cirq-core/cirq/ops/linear_combinations.py:845]: the change got lost

File: cirq-core/cirq/ops/linear_combinations_test.py (hunk 5 above)
Comment (pavoljuhas) [reply cirq-core/cirq/ops/linear_combinations_test.py:1164]: the update is not in effect

Comment (TarunSinghania) [reply cirq-core/cirq/ops/linear_combinations_test.py]: Thanks for your input, the above was happening as I was using amend on my local system after pushing commits to remote. This was creating diversions and required proper merge/rebase. I will be adding new commits from now on.

I have corrected the above by adding a new [commit](https://github.com/quantumlib/Cirq/pull/6019/commits/b12c60d7b27172ea20fc984b3f15eacbb7611d18)

File: cirq-core/cirq/ops/linear_combinations_test.py (hunk 6 above)
Comment (TarunSinghania) [reply cirq-core/cirq/ops/linear_combinations_test.py]: I have removed the trailing blanks, but even when I add a white space and do git diff --check master...HEAD it doesn't show any complaints. I haven't dug into it much to see if I am missing something here.
For now I think checks on ./check/all should help

**Comment by TarunSinghania** [comment]
> Hi Tarun, when updating the PR, please add new changes as new commits on top of your branch rather than amending old commits, which may result in complicated commit graphs with duplicate original and amended commits. It is also not necessary to merge-in the mainline master in every update - provided the master does not touch files worked on in this PR.
> 
> It is not a big deal as in the end we squash each PR to a single commit on master, but while the PR is open it is easier to follow its progress with a simple commit history.

Thanks for the comment. I am new to using git, I thought amending a commit would edit the last commit. 
Now I understand the behavior, that it only amends the last commit  on the local repository (on my laptop). 
The commit on the remote (forked repository) is already written. 
e.g
commit-1 -> commit-2 -> push to remote 
Now if I amend commit-2 locally, the history locally will be commit1->commit3. Locally the branch would be one commit (commit-2) behind the main branch 
We will have two diverging histories for the same branch, which we will have to either: 
1. Merge 2. Rebase
Using github desktop and generating a Pull request automatically uses the merge option and led to complicated commit graphs with duplicate original and amended commits. 

I will be adding in new commits on top of my branch (which is main branch of my local repository for this issue, but will create separate branches from now on).

File: cirq-core/cirq/ops/linear_combinations_test.py [hunk 7]
```diff
@@ -1156,6 +1156,23 @@ def test_pauli_sum_pow():
     for psum in [psum1, psum2, psum3, psum4]:
         assert cirq.approx_eq(psum**0, identity)
 
+    # tests for exponents greater than two for both even and odd
+    psum5 = cirq.Z(q0) * cirq.Z(q1) + cirq.Z(q2) + cirq.Z(q3)
+    correctresult = psum5.copy()
+    for e in range(1,9):
+        powfunctionresult = psum5**e
+        assert correctresult == powfunctionresult
```
Comment (pavoljuhas) [inline cirq-core/cirq/ops/linear_combinations_test.py:1164]: Nit: I suggest we do here without the local variable:
```
assert correctresult == psum5**e
```

**Review by pavoljuhas — COMMENTED** [review]
Thank you for the updates Tarun.
LGTM with 2 small suggestions, also please fix the formatting issues reported by the CI check.
Otherwise it is good to go.

File: cirq-core/cirq/ops/linear_combinations_test.py (hunk 7 above)
Comment (TarunSinghania) [reply cirq-core/cirq/ops/linear_combinations_test.py:1164]: Sure, no need for having a local variable

File: cirq-core/cirq/ops/linear_combinations_test.py [hunk 8]
```diff
@@ -1156,6 +1156,23 @@ def test_pauli_sum_pow():
...
+        correctresult *= psum5
+
+    psum6 = cirq.X(q0) * cirq.Y(q1) + cirq.Z(q2) + cirq.X(q3)
+    assert psum6 * psum6 * psum6 * psum6 * psum6 * psum6 * psum6 * psum6 == psum6**8
+
+    # test to ensure pow doesn't make any change to the original value
+    psum7 = cirq.X(q0) * cirq.Y(q1) + cirq.Z(q2)
+    psum7copy = psum7.copy()
+    psum7raised5 = psum7**5
+    assert psum7raised5 == psum7 * psum7 * psum7 * psum7 * psum7
```
[the remark this answers is not in the document]
Comment (TarunSinghania) [reply cirq-core/cirq/ops/linear_combinations_test.py:1174]: done

File: cirq-core/cirq/ops/linear_combinations.py (hunk 4 above)
Comment (TarunSinghania) [reply cirq-core/cirq/ops/linear_combinations.py]: Corrected

**Review by pavoljuhas — APPROVED** [review]
LGTM

**Review by tanujkhattar — APPROVED** [review]

**Pull request merged**